### PR TITLE
fix(carousel): Animation Callback in Angular 1.4

### PR DESCRIPTION
**Pull request by idealeric**
$animate no longer triggers events on the element itself, but rather has its own interface for managing these callbacks.

I wasn't sure what the protocal for supporting angular 1.3 vs 1.4 was, so I just put the version check in. 
Grunt passed locally and also worked in my manual testing.

Let me know. :)

### CODE REVIEW

**Comment by wesleycho** [comment]
See the discussion in #3811 for guidance on the preferred implementation.

File: src/carousel/carousel.js [hunk 1]
```diff
@@ -40,9 +40,18 @@ angular.module('ui.bootstrap.carousel', [])
...
       slide.$element.data(SLIDE_DIRECTION, slide.direction);
       $scope.$currentTransition = true;
-      slide.$element.one('$animate:close', function closeFn() {
-        $scope.$currentTransition = null;
-      });
+      if ( angular.version.minor >= 4 ) {
+        $animate.on('addClass', slide.$element, function closeFn14(aniElem, phase) {
+          if ( phase === 'close' ) {
+            $scope.$currentTransition = null;
+            $animate.off('addClass', aniElem);
```
Comment (chrisirhc) [inline src/carousel/carousel.js]: Shouldn't this be `closeFn14` ?

**Comment by chrisirhc** [comment]
Did a review. This looks good. I thought at first that we need to check the major version of angular but I realized that if Angular 2 is used, it would probably be a breaking change anyways.

**Comment by wesleycho** [comment]
Ideally, shouldn't we migrate over to $animateCss conditionally through usage of $injector similar to the suggestion that was made in the aforementioned issue? It contains explicit support for adding and removing classes, which I would be concerned about code fragility with ngAnimate support in abusing $animate to get 1.4 support given that it has been stated that $animateCss should be used for anything related to CSS animations (and that CSS animations otherwise would not be supported).

If this patch is to make it in, I would want to refactor it asap to change over to use $animateCss in the 1.4+ situation, since it could potentially be dangerous.

**Comment by chrisirhc** [comment]
You're right, we still need to migrate over to `$animateCss` to do this animation the correct way.
However, I think this change is still needed even if we switch over to using `$animateCss` and this could be the bare minimum to get things working in 1.4 (I haven't tested to verify this). 

I'm not at all opposed to getting the $animateCss refactor in. It's just that if this is the bare minimum, we shouldn't block merging this since we can always do the $animateCss refactor later and things would still be working for a while.

File: src/carousel/carousel.js [hunk 2]
```diff
@@ -40,9 +40,18 @@ angular.module('ui.bootstrap.carousel', [])
       slide.$element) {
       slide.$element.data(SLIDE_DIRECTION, slide.direction);
       $scope.$currentTransition = true;
-      slide.$element.one('$animate:close', function closeFn() {
-        $scope.$currentTransition = null;
-      });
+      if ( angular.version.minor >= 4 ) {
```
Comment (chrisirhc) [inline src/carousel/carousel.js]: Alternatively, test for existence of `$animate.on`.

**Comment by idealeric** [comment]
I made these minor changes. :)

Also i'm with @chrisirhc. From my testing, the use of $animateCss is actually more or less equivalent to the use of .addClass currently. I would not be surprised if .addClass uses animateCss under the sheets anyway (the docs said as much, but I did not bother reading and angular source code).

**Comment by wesleycho** [comment]
My comment about $animateCss is mainly due to the official stance on the ngAnimate api - in the future, it is conceivable that $animate will make a breaking change with regards to how it handles animation involving classes with $animate.

That said, I am not against getting the change in without $animateCss support, but it makes changing to use $animateCss for 1.4+ more urgent whenever possible to avoid breakage in the future.

**Comment by wesleycho** [comment]
I did a build based off of this PR, and it seems like the animation is a little borked - check out the Plunker [here](http://plnkr.co/edit/XBChPC6SJPvkUVgNdiwk?p=preview) to see the UI glitch.

**Comment by chrisirhc** [comment]
Hm.. indeed, in the Plunker, it looks like the classes from `$animate.addClass` isn't being applied. Looks like we'll need $animateCss after all.

**Pull request closed without merging**